lib/semantics/statements.cc

```diff
@@ -0,0 +1,97 @@
+#include "statements.h"
+#include "semantics.h"
+#include "../common/indirection.h"
+#include "../common/Fortran.h"
+#include "../evaluate/expression.h"
+#include "../parser/parse-tree.h"
+#include <optional>
+#include <variant>
+
+namespace Fortran::semantics {
+
+class StatementContext {
+public:
+  explicit StatementContext(
+      SemanticsContext &c, parser::CharBlock at = parser::CharBlock{})
+    : context_{c}, messages_{at, &c.messages()} {}
+
+  bool operator==(const StatementContext &x) const { return this == &x; }
+  bool operator!=(const StatementContext &x) const { return this != &x; }
+
+  void Analyze(const parser::StopStmt &stmt) {
+    const auto &sc{std::get<std::optional<parser::StopCode>>(stmt.t)};
+    const auto &sle{std::get<std::optional<parser::ScalarLogicalExpr>>(stmt.t)};
+
+    if (sc) {
+      const parser::CharBlock &source{sc.value().v.thing.source};
+      const auto &type{sc.value().v.thing.typedExpr->v.GetType()};
+
+      if (!type) {
+        messages_.Say(source,
+          "type information not ready for STOP code"_err_en_US);
+      } else {
+        common::TypeCategory tc{type.value().category};
+
+        if (tc == common::TypeCategory::Integer) {
+          int kind{type.value().kind};
+
+          // C1170 default kind
+          if (context_.defaultKinds().GetDefaultKind(tc) != kind) {
+            messages_.Say(source,
+              "integer parameter of STOP not of default kind"_err_en_US);
+          }
+        } else if (tc != common::TypeCategory::Character) {
+          messages_.Say(source,
+            "STOP patameter neither of INTEGER or CHARACTER type"_err_en_US);
+        }
+      }
+    }
+    if (sle) {
+      const parser::CharBlock &source{sle.value().thing.thing.value().source};
+      const auto &type{sle.value().thing.thing.value().typedExpr->v.GetType()};
+
+      if (!type) {
+        messages_.Say(source,
+          "type information not ready for QUIET parameter of the STOP statement"_err_en_US);
+      } else {
+        common::TypeCategory tc{type.value().category};
+        if (tc != common::TypeCategory::Logical) {
+          messages_.Say(source,
+            "QUIET patameter of the STOP statement is not of LOGICAL type"_err_en_US);
+        }
+      }
+    }
```

Comment:
Do you want to check that either `sc` or `sle` has a value but not both and not neither? (That's implied by R1162.)

Reply:
Do you mean `ERROR STOP 123, QUIET = .true.` (where `sc = 123`, `sle = .true.`) is illegal?